Foundation/NSKeyedUnarchiver.swift

```diff
@@ -0,0 +1,970 @@
+        }
+
+        return _currentDecodingContext.dict[unwrappedKey!] as? T
+    }
+    
+    /**
+        Dereferences, but does not decode, an object reference
+     */
+    private func _dereferenceObjectReference(unwrappedObjectRef: CFKeyedArchiverUID) -> Any? {
+        let uid = Int(objectRefGetValue(unwrappedObjectRef))
+        
+        guard uid < self._objects.count else {
+            return nil
+        }
+
+        return self._objects[uid]
+    }
+    
+    public override var systemVersion: UInt32 {
+        return NSKeyedArchiverSystemVersion
+    }
+    
+    public override var allowsKeyedCoding: Bool {
+        get {
+            return true
+        }
+    }
+    
+    private func _validateStillDecoding() -> Bool {
+        if self._flags.contains(UnarchiverFlags.FinishedDecoding) {
+            fatalError("Decoder already finished")
+        }
+        
+        return true
+    }
+    
+    private static func _supportsSecureCoding(clsv : AnyClass) -> Bool {
+        if let secureCodable = clsv as? NSSecureCoding.Type {
+            return secureCodable.supportsSecureCoding()
+        }
+        
+        return false
+    }
+    
+    // FIXME is there a better way to do this with Swift stdlib?
+    private static func _classIsKindOfClass(assertedClass : AnyClass, _ allowedClass : AnyClass) -> Bool {
+        var superClass : AnyClass? = assertedClass
+        
+        repeat {
+            if superClass == allowedClass {
+                return true
+            }
+            
+            superClass = _getSuperclass(superClass!)
+        } while superClass != nil
+        
+        return false
+    }
+    
+    private func _isClassAllowed(assertedClass: AnyClass?, allowedClasses: [AnyClass]?) -> Bool {
+        if assertedClass == nil {
+            return false
+        }
+        
+        if _flags.contains(UnarchiverFlags.RequiresSecureCoding) {
+            if let unwrappedAllowedClasses = allowedClasses {
+                if unwrappedAllowedClasses.contains({NSKeyedUnarchiver._classIsKindOfClass(assertedClass!, $0)}) {
+                    return true
+                }
+            }
+            
+            fatalError("Value was of unexpected class \(assertedClass!)")
+        } else {
+            return true
+        }
+    }
+   
+    /**
+        Validate a dictionary with class type information, mapping to a class if allowed
+     */ 
+    private func _validateAndMapClassDictionary(classDict: Dictionary<String, Any>?,
+                                                allowedClasses: [AnyClass]?,
+                                                inout classToConstruct: AnyClass?) -> Bool {
+        classToConstruct = nil
+        
+        func _classForClassName(codedName: String) -> AnyClass? {
+            var aClass : AnyClass?
+            
+            aClass = classForClassName(codedName)
+            if aClass == nil {
+                aClass = NSKeyedUnarchiver.classForClassName(codedName)
```

Comment:
I think we should probably have the base Foundation types stuffed in the mappings backing this so that we can avoid the call to NSClassFromString as long as possible since that will require sign-off from the runtime team for us accessing their internal functions